### ISSUE TITLE
[FEATURE_REQUEST] flag on findOrCreate result to answer: "Was it found or created?"

### ISSUE THREAD

**Issue by timestep** [balderdashy/waterline#294]
https://github.com/balderdashy/sails/issues/1472

Is there a way to determine whether the record was found or created in the callback?

Right now, it seems if you pass in the record id and the record you want found or created, it simply returns the new record whether it was created or not. It does't tell you if it found it or not, and I kinda of wanted a work flow that was sensitive to this.

I know the work around is simply to do a findOne before, but this would be a nice feature.

**Comment by ksaitor** [comment]
@itslenny +1 for `updateOrCreate`.

**Comment by anasyb** [comment]
+1 for `updateOrCreate` 
+1 for found or created

**Comment by dmarcelino** [comment]
`updateOrCreate` has its own issue: #790, please add any upvotes there.

Regarding this specific feature, I like what @particlebanana suggested on https://github.com/balderdashy/waterline/issues/294#issuecomment-36778354: 
> I think it should be an array of objects with and id or whatever your PK is and a flag like you mentioned. So a `findOrCreate` would be an array with one value and a `findOrCreateEach` would be an array with all the values.

**Comment by Globegitter** [comment]
@ksaitor @itslenny for `updateOrCreate` check out https://github.com/balderdashy/waterline/pull/940

I am generally for the idea but against the third callback. It should be on the results object itself. Otherwise it wouldn't work properly with future syntax such as `async/await`.

**Comment by minecrafter** [comment]
Just ran into this very issue myself. For now, I'm likely to work around it by checking the creation date of the returned object to make sure it is within a certain delta of time but feel this is too hacky.

:+1: for this

**Comment by dmarcelino** [comment]
> I am generally for the idea but against the third callback. It should be on the results object itself. Otherwise it wouldn't work properly with future syntax such as async/await.

Changing the results object (which is currently an array) could prove troublesome for 2 reasons:
* It won't be backwards compatible;
* It will make `.findOrCreate()` API different than other methods.

Unless we return an array with its prototype extended to provide the flags...

**Comment by dmarcelino** [comment]
> For now, will need to use non-atomic find/create...

@minecrafter, currently `.findOrCreate` is only atomic for adapters that implement it. For adapters that don't implement it will never be atomic until transactions (#755) are implemented.

**Comment by sailsbot** [comment]
Thanks for posting, @timestep.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
